integrations/azure-devops/azure_devops/http/http_client.py

```diff
@@ -0,0 +1,150 @@
+        # Username isn't required in basic auth to Azure Devops
+        self._auth = BasicAuth("", personal_access_token)
+        self._timeout = Timeout(REQUEST_TIMEOUT_SECONDS)
+        self._follow_redirects = ALLOW_REDIRECTS
+
+    @staticmethod
+    def _parse_response_values(response: Response) -> list[dict[Any, Any]]:
+        return response.json()["value"]
+
+    def send_sync_get_request(
+        self, url: str, params: Optional[dict[str, Any]] = None
+    ) -> Response:
+        try:
+            response = httpx.get(
+                url,
+                params=params,
+                auth=self._auth,
+                timeout=self._timeout,
+                follow_redirects=self._follow_redirects,
+            )
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            if response.status_code == 401:
+                logger.error(
+                    f"Couldn't access url {url} . Make sure the PAT (Personal Access Token) is valid!"
+                )
+            else:
+                logger.debug(
+                    f"Request with bad status code {response.status_code}: GET to url {url}: {str(e)}"
+                )
+                raise e
+        except Exception as e:
+            logger.error(f"Couldn't send get request to url {url}: {str(e)}")
+            raise e
+        return response
+
+    async def send_get_request(
+        self,
+        url: str,
+        params: Optional[dict[str, Any]] = None,
+        headers: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        return await self._send_request(
+            method="get", url=url, params=params, headers=headers
+        )
+
+    async def send_post_request(
+        self,
+        url: str,
+        params: Optional[dict[str, Any]] = None,
+        data: Optional[str | dict[Any, Any]] = None,
+        headers: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        return await self._send_request(
+            method="post", params=params, url=url, data=data, headers=headers
+        )
+
+    async def _send_request(
+        self,
+        method: str,
+        url: str,
+        data: Optional[Any] = None,
+        params: Optional[dict[str, Any]] = None,
+        headers: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        try:
+            response = await self._async_client.request(
+                method=method,
+                url=url,
+                data=data,
+                params=params,
+                auth=self._auth,
+                follow_redirects=self._follow_redirects,
+                timeout=self._timeout,
+                headers=headers,
+            )
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            if response.status_code == 401:
+                logger.error(
+                    f"Couldn't access url {url} . Make sure the PAT (Personal Access Token) is valid!"
+                )
```

Comment:
Why would it try to access something that it does not permissions for?
If there is no reason then please raise the error so we wont create inconsistent state

Reply:
The PAT can become stale, thats the reason that I put this log. There shouldn't be any other case leading to this (providing the customer set the proper permissions to the PAT)